conn.go

```diff
@@ -939,21 +939,23 @@ func (c *Conn) do(d *connDeadline, write func(time.Time, int32) error, read func
 		return err
 	}
 
-	deadline, size, lock, err := c.waitResponse(d, id)
-	if err != nil {
-		return err
-	}
+	if c.requiredAcks != 0 {
```

Comment:
Don't we have to check that this is a produce request? `do` is used for all operations, but "required acks" is only employed when send messages isn't it?

Reply:
I totally agree with you but the `waitResponse` is done at connection level (so in the conn.go file) where as far as I can see from the code we don't have information if it's a producer or a consumer request (which is at higher level). We should do a refactoring for moving the waiting for the response at the upper layer.
Is that really a problem? I mean, the conn is always initialized with -1 as a default value so other requests will work out of the box (and for consumer we cannot change it).
Btw if you think there is a simple way to check for producer request, feedback are really appreciated :-)

Reply:
Just adding that if the conn is there for representing a connection to the broker, the `requiredAcks` should not be an information at this so low level ... but it was already there.

Reply:
We could thing to add a new `waitResponse bool` parameter for the `do` function passed by the `writeOperation` and `readOperation` used by higher level protocol operation for which we know if it's needed to wait a response or not.

Reply:
I think we're pretty much forced to handle it at this level since it's core to how the kafka protocol works (some requests have responses, some don't, currently we only support the former), and we use the `Conn` abstraction for all interactions with kafka brokers.

We don't know how the connection is being handled, but we do know what kafka API call is being made, so I think we can figure out whether the "required ack" applies or not based on that, can't we?

It's OK to modify internal APIs to support new use cases as well, if we need to pass more info down to this method then let's do it.

Reply:
> We don't know how the connection is being handled, but we do know what kafka API call is being made, so I think we can figure out whether the "required ack" applies or not based on that, can't we?

Which information inside conn.go provides the information about which API is called. Isn't it an information of higher level?

Reply:
The entry point for producing messages is here https://github.com/segmentio/kafka-go/blob/master/conn.go#L828

So this is where we have most of the context for deciding whether waiting for a ack is required or not.

I think the simplest is going to be copying the content of the `do` method into `WriteCompressedMessages`, in replacement for the call to `writeOperation`, and based on whether "required acks" are set or not we wait for the response.

What do you think about this?

Reply:
@ppatierno do you still have interest in pushing this through the finish line?